File: src/main/java/me/miki/shindo/management/mods/impl/BlockOverlayMod.java
package me.miki.shindo.management.mods.impl;

import me.miki.shindo.Shindo;
import me.miki.shindo.management.color.AccentColor;
import me.miki.shindo.management.event.EventTarget;
import me.miki.shindo.management.event.impl.EventBlockHighlightRender;
import me.miki.shindo.management.language.TranslateText;
import me.miki.shindo.management.mods.Mod;
import me.miki.shindo.management.mods.ModCategory;
import me.miki.shindo.utils.ColorUtils;
import me.miki.shindo.utils.Render3DUtils;
import me.miki.shindo.utils.TimerUtils;
import me.miki.shindo.utils.animation.simple.SimpleAnimation;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderGlobal;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.inventory.IInventory;
import net.minecraft.item.ItemStack;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.BlockPos;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.world.WorldSettings;
import org.lwjgl.opengl.GL11;

import java.awt.*;

import me.miki.shindo.management.settings.config.Property;
import me.miki.shindo.management.settings.config.PropertyType;
public class BlockOverlayMod extends Mod {

    private final SimpleAnimation[] simpleAnimation = {new SimpleAnimation(0.0F), new SimpleAnimation(0.0F), new SimpleAnimation(0.0F), new SimpleAnimation(0.0F), new SimpleAnimation(0.0F), new SimpleAnimation(0.0F)};

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ANIMATION)
    private boolean animationSetting = false;

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.FILL)
    private boolean fillSetting = true;

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.OUTLINE)
    private boolean outlineSetting = true;

    @Property(type = PropertyType.NUMBER, translate = TranslateText.FILL_ALPHA, min = 0, max = 1.0, current = 0.15)
    private double fillAlphaSetting = 0.15;

    @Property(type = PropertyType.NUMBER, translate = TranslateText.OUTLINE_ALPHA, min = 0, max = 1.0, current = 0.15)
    private double outlineAlphaSetting = 0.15;

    @Property(type = PropertyType.NUMBER, translate = TranslateText.OUTLINE_WIDTH, min = 1, max = 10, current = 4)
    private double outlineWidthSetting = 4;

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.DEPTH)
    private boolean depthSetting = false;

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.CUSTOM_COLOR)
    private boolean customColorSetting = false;

    @Property(type = PropertyType.COLOR, translate = TranslateText.FILL_COLOR)
    private Color fillColorSetting = Color.RED;

    @Property(type = PropertyType.COLOR, translate = TranslateText.OUTLINE_COLOR)
    private Color outlineColorSetting = Color.RED;

    protected AxisAlignedBB currentBB;
    protected AxisAlignedBB slideBB;
    protected TimerUtils timer = new TimerUtils();

    public BlockOverlayMod() {
        super(TranslateText.BLOCK_OVERLAY, TranslateText.BLOCK_OVERLAY_DESCRIPTION, ModCategory.RENDER, "blockoutline");
    }

    @EventTarget
    public void onBlockHighlightRender(EventBlockHighlightRender event) {

        AccentColor currentColor = Shindo.getInstance().getColorManager().getCurrentColor();

        event.setCancelled(true);

        if (!canRender(event.getObjectMouseOver())) {
            return;
        }

        GlStateManager.enableBlend();
        GlStateManager.tryBlendFuncSeparate(770, 771, 1, 0);

        if (depthSetting) {
            GlStateManager.disableDepth();
        }

        GlStateManager.disableTexture2D();

        GlStateManager.depthMask(false);
        BlockPos blockpos = event.getObjectMouseOver().getBlockPos();
        Block block = mc.theWorld.getBlockState(blockpos).getBlock();

        if (block.getMaterial() != Material.air && mc.theWorld.getWorldBorder().contains(blockpos)) {

            block.setBlockBoundsBasedOnState(mc.theWorld, blockpos);

            double x = mc.getRenderViewEntity().lastTickPosX
                    + (mc.getRenderViewEntity().posX - mc.getRenderViewEntity().lastTickPosX) * (double) event.getPartialTicks();
            double y = mc.getRenderViewEntity().lastTickPosY
                    + (mc.getRenderViewEntity().posY - mc.getRenderViewEntity().lastTickPosY) * (double) event.getPartialTicks();
            double z = mc.getRenderViewEntity().lastTickPosZ
                    + (mc.getRenderViewEntity().posZ - mc.getRenderViewEntity().lastTickPosZ) * (double) event.getPartialTicks();

            AxisAlignedBB selectedBox = block.getSelectedBoundingBox(mc.theWorld, blockpos);

            if (animationSetting) {

                if (!selectedBox.equals(currentBB)) {
                    slideBB = currentBB;
                    currentBB = selectedBox;
                }

                AxisAlignedBB slide;

                if ((slide = slideBB) != null) {

                    simpleAnimation[0].setAnimation((float) (slide.minX + (selectedBox.minX - slide.minX)), 24);
                    simpleAnimation[1].setAnimation((float) (slide.minY + (selectedBox.minY - slide.minY)), 24);
                    simpleAnimation[2].setAnimation((float) (slide.minZ + (selectedBox.minZ - slide.minZ)), 24);
                    simpleAnimation[3].setAnimation((float) (slide.maxX + (selectedBox.maxX - slide.maxX)), 24);
                    simpleAnimation[4].setAnimation((float) (slide.maxY + (selectedBox.maxY - slide.maxY)), 24);
                    simpleAnimation[5].setAnimation((float) (slide.maxZ + (selectedBox.maxZ - slide.maxZ)), 24);

                    AxisAlignedBB renderBB = new AxisAlignedBB(
                            simpleAnimation[0].getValue() - 0.01,
                            simpleAnimation[1].getValue() - 0.01,
                            simpleAnimation[2].getValue() - 0.01,
                            simpleAnimation[3].getValue() + 0.01,
                            simpleAnimation[4].getValue() + 0.01,
                            simpleAnimation[5].getValue() + 0.01
                    );

                    if (fillSetting) {
                        ColorUtils.setColor(customColorSetting ? fillColorSetting.getRGB() : currentColor.getInterpolateColor().getRGB(), (float) fillAlphaSetting);
                        Render3DUtils.drawFillBox(interpolateAxis(renderBB));
                    }

                    if (outlineSetting) {
                        ColorUtils.setColor(customColorSetting ? outlineColorSetting.getRGB() : currentColor.getInterpolateColor().getRGB(), (float) outlineAlphaSetting);
                        GL11.glLineWidth((float) outlineWidthSetting);
                        RenderGlobal.drawSelectionBoundingBox(interpolateAxis(renderBB));
                    }
                }
            } else {

                selectedBox = selectedBox.expand(0.0020000000949949026D, 0.0020000000949949026D, 0.0020000000949949026D).offset(-x, -y, -z);

                if (fillSetting) {
                    ColorUtils.setColor(customColorSetting ? fillColorSetting.getRGB() : currentColor.getInterpolateColor().getRGB(), (float) fillAlphaSetting);
                    Render3DUtils.drawFillBox(selectedBox);
                }

                if (outlineSetting) {
                    ColorUtils.setColor(customColorSetting ? outlineColorSetting.getRGB() : currentColor.getInterpolateColor().getRGB(), (float) outlineAlphaSetting);
                    GL11.glLineWidth((float) outlineWidthSetting);
                    RenderGlobal.drawSelectionBoundingBox(selectedBox);
                }
            }
        }

        GlStateManager.depthMask(true);
        GlStateManager.enableTexture2D();

        GlStateManager.disableBlend();

        if (depthSetting) {
            GlStateManager.enableDepth();
        }

        GL11.glLineWidth(2);
    }

    private boolean canRender(MovingObjectPosition movingObjectPositionIn) {

        Entity entity = mc.getRenderViewEntity();
        boolean result = entity instanceof EntityPlayer && !mc.gameSettings.hideGUI;

        if (result && !((EntityPlayer) entity).capabilities.allowEdit) {
            ItemStack itemstack = ((EntityPlayer) entity).getCurrentEquippedItem();

            if (mc.objectMouseOver != null && mc.objectMouseOver.typeOfHit == MovingObjectPosition.MovingObjectType.BLOCK) {
                BlockPos selectedBlock = mc.objectMouseOver.getBlockPos();
                Block block = mc.theWorld.getBlockState(selectedBlock).getBlock();

                if (mc.playerController.getCurrentGameType() == WorldSettings.GameType.SPECTATOR) {
                    result = block.hasTileEntity() && mc.theWorld.getTileEntity(selectedBlock) instanceof IInventory;
                } else {
                    result = itemstack != null && (itemstack.canDestroy(block) || itemstack.canPlaceOn(block));
                }
            }
        }

        result = result && movingObjectPositionIn.typeOfHit == MovingObjectPosition.MovingObjectType.BLOCK;

        return result;
    }

    public AxisAlignedBB interpolateAxis(AxisAlignedBB bb) {
        return new AxisAlignedBB(
                bb.minX - mc.getRenderManager().viewerPosX,
                bb.minY - mc.getRenderManager().viewerPosY,
                bb.minZ - mc.getRenderManager().viewerPosZ,
                bb.maxX - mc.getRenderManager().viewerPosX,
                bb.maxY - mc.getRenderManager().viewerPosY,
                bb.maxZ - mc.getRenderManager().viewerPosZ);
    }
}

File: src/main/kotlin/me/miki/shindo/management/mods/impl/BlockOverlayMod.kt
package me.miki.shindo.management.mods.impl

import me.miki.extensions.ui.animation.setAnimation
import me.miki.shindo.Shindo
import me.miki.shindo.management.event.EventTarget
import me.miki.shindo.management.event.impl.EventBlockHighlightRender
import me.miki.shindo.management.language.TranslateText
import me.miki.shindo.management.mods.Mod
import me.miki.shindo.management.mods.ModCategory
import me.miki.shindo.management.nanovg.font.Lucide
import me.miki.shindo.management.nanovg.font.Shinconic
import me.miki.shindo.management.settings.config.Property
import me.miki.shindo.management.settings.config.PropertyType
import me.miki.shindo.ui.animation.v2.value.SimpleAnimation
import me.miki.shindo.utils.ColorUtils.setColor
import me.miki.shindo.utils.Render3DUtils.drawFillBox
import me.miki.shindo.utils.TimerUtils
import net.minecraft.block.material.Material
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderGlobal
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.inventory.IInventory
import net.minecraft.util.AxisAlignedBB
import net.minecraft.util.MovingObjectPosition
import net.minecraft.util.MovingObjectPosition.MovingObjectType
import net.minecraft.world.WorldSettings
import org.lwjgl.opengl.GL11
import java.awt.Color

open class BlockOverlayMod :
    Mod(
        TranslateText.BLOCK_OVERLAY,
        TranslateText.BLOCK_OVERLAY_DESCRIPTION,
        ModCategory.RENDER,
        Shinconic.MOD_BLOCK_OVERLAY,
        "blockoutline",
    ) {
    private val simpleAnimation =
        arrayOf<SimpleAnimation?>(
            SimpleAnimation(0.0f),
            SimpleAnimation(0.0f),
            SimpleAnimation(0.0f),
            SimpleAnimation(0.0f),
            SimpleAnimation(0.0f),
            SimpleAnimation(0.0f),
        )

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ANIMATION)
    private val animationSetting = false

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.FILL)
    private val fillSetting = true

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.OUTLINE)
    private val outlineSetting = true

    @Property(type = PropertyType.NUMBER, translate = TranslateText.FILL_ALPHA, min = 0.0, max = 1.0, current = 0.15)
    private val fillAlphaSetting = 0.15

    @Property(type = PropertyType.NUMBER, translate = TranslateText.OUTLINE_ALPHA, min = 0.0, max = 1.0, current = 0.15)
    private val outlineAlphaSetting = 0.15

    @Property(type = PropertyType.NUMBER, translate = TranslateText.OUTLINE_WIDTH, min = 1.0, max = 10.0, current = 4.0)
    private val outlineWidthSetting = 4.0

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.DEPTH)
    private val depthSetting = false

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.CUSTOM_COLOR)
    private val customColorSetting = false

    @Property(type = PropertyType.COLOR, translate = TranslateText.FILL_COLOR)
    private val fillColorSetting: Color = Color.RED

    @Property(type = PropertyType.COLOR, translate = TranslateText.OUTLINE_COLOR)
    private val outlineColorSetting: Color = Color.RED

    private var currentBB: AxisAlignedBB? = null
    private var slideBB: AxisAlignedBB? = null
    protected var timer: TimerUtils = TimerUtils()

    @EventTarget
    fun onBlockHighlightRender(event: EventBlockHighlightRender) {
        val currentColor = Shindo.getInstance().getColorManager().getCurrentColor()

        event.setCancelled(true)

        if (!canRender(event.objectMouseOver)) {
            return
        }

        GlStateManager.enableBlend()
        GlStateManager.tryBlendFuncSeparate(770, 771, 1, 0)

        if (depthSetting) {
            GlStateManager.disableDepth()
        }

        GlStateManager.disableTexture2D()

        GlStateManager.depthMask(false)
        val blockpos = event.objectMouseOver.blockPos
        val block = mc.theWorld.getBlockState(blockpos).block

        if (block.material !== Material.air && mc.theWorld.worldBorder.contains(blockpos)) {
            block.setBlockBoundsBasedOnState(mc.theWorld, blockpos)

            val x = (
                mc.renderViewEntity.lastTickPosX +
                    (mc.renderViewEntity.posX - mc.renderViewEntity.lastTickPosX) *
                    event
                        .getPartialTicks()
                        .toDouble()
            )
            val y = (
                mc.renderViewEntity.lastTickPosY +
                    (mc.renderViewEntity.posY - mc.renderViewEntity.lastTickPosY) *
                    event
                        .getPartialTicks()
                        .toDouble()
            )
            val z = (
                mc.renderViewEntity.lastTickPosZ +
                    (mc.renderViewEntity.posZ - mc.renderViewEntity.lastTickPosZ) *
                    event
                        .getPartialTicks()
                        .toDouble()
            )

            var selectedBox = block.getSelectedBoundingBox(mc.theWorld, blockpos)

            if (animationSetting) {
                if (selectedBox != currentBB) {
                    slideBB = currentBB
                    currentBB = selectedBox
                }

                val slide: AxisAlignedBB?

                if ((slideBB.also { slide = it }) != null) {
                    simpleAnimation[0]!!.setAnimation((slide!!.minX + (selectedBox.minX - slide.minX)).toFloat(), 24)
                    simpleAnimation[1]!!.setAnimation((slide.minY + (selectedBox.minY - slide.minY)).toFloat(), 24)
                    simpleAnimation[2]!!.setAnimation((slide.minZ + (selectedBox.minZ - slide.minZ)).toFloat(), 24)
                    simpleAnimation[3]!!.setAnimation((slide.maxX + (selectedBox.maxX - slide.maxX)).toFloat(), 24)
                    simpleAnimation[4]!!.setAnimation((slide.maxY + (selectedBox.maxY - slide.maxY)).toFloat(), 24)
                    simpleAnimation[5]!!.setAnimation((slide.maxZ + (selectedBox.maxZ - slide.maxZ)).toFloat(), 24)

                    val renderBB =
                        AxisAlignedBB(
                            simpleAnimation[0]!!.getValue() - 0.01,
                            simpleAnimation[1]!!.getValue() - 0.01,
                            simpleAnimation[2]!!.getValue() - 0.01,
                            simpleAnimation[3]!!.getValue() + 0.01,
                            simpleAnimation[4]!!.getValue() + 0.01,
                            simpleAnimation[5]!!.getValue() + 0.01,
                        )

                    if (fillSetting) {
                        setColor(
                            if (customColorSetting) {
                                fillColorSetting.rgb
                            } else {
                                currentColor
                                    .getInterpolateColor()
                                    .rgb
                            },
                            fillAlphaSetting.toFloat(),
                        )
                        drawFillBox(interpolateAxis(renderBB))
                    }

                    if (outlineSetting) {
                        setColor(
                            if (customColorSetting) {
                                outlineColorSetting.rgb
                            } else {
                                currentColor
                                    .getInterpolateColor()
                                    .rgb
                            },
                            outlineAlphaSetting.toFloat(),
                        )
                        GL11.glLineWidth(outlineWidthSetting.toFloat())
                        RenderGlobal.drawSelectionBoundingBox(interpolateAxis(renderBB))
                    }
                }
            } else {
                selectedBox =
                    selectedBox
                        .expand(0.0020000000949949026, 0.0020000000949949026, 0.0020000000949949026)
                        .offset(-x, -y, -z)

                if (fillSetting) {
                    setColor(
                        if (customColorSetting) {
                            fillColorSetting.rgb
                        } else {
                            currentColor
                                .getInterpolateColor()
                                .rgb
                        },
                        fillAlphaSetting.toFloat(),
                    )
                    drawFillBox(selectedBox)
                }

                if (outlineSetting) {
                    setColor(
                        if (customColorSetting) {
                            outlineColorSetting.rgb
                        } else {
                            currentColor
                                .getInterpolateColor()
                                .rgb
                        },
                        outlineAlphaSetting.toFloat(),
                    )
                    GL11.glLineWidth(outlineWidthSetting.toFloat())
                    RenderGlobal.drawSelectionBoundingBox(selectedBox)
                }
            }
        }

        GlStateManager.depthMask(true)
        GlStateManager.enableTexture2D()

        GlStateManager.disableBlend()

        if (depthSetting) {
            GlStateManager.enableDepth()
        }

        GL11.glLineWidth(2f)
    }

    private fun canRender(movingObjectPositionIn: MovingObjectPosition): Boolean {
        val entity = mc.renderViewEntity
        var result = entity is EntityPlayer && !mc.gameSettings.hideGUI

        if (result && !(entity as EntityPlayer).capabilities.allowEdit) {
            val itemstack = entity.currentEquippedItem

            if (mc.objectMouseOver != null && mc.objectMouseOver.typeOfHit == MovingObjectType.BLOCK) {
                val selectedBlock = mc.objectMouseOver.blockPos
                val block = mc.theWorld.getBlockState(selectedBlock).block

                result =
                    if (mc.playerController.currentGameType == WorldSettings.GameType.SPECTATOR) {
                        block.hasTileEntity() && mc.theWorld.getTileEntity(selectedBlock) is IInventory
                    } else {
                        itemstack != null && (itemstack.canDestroy(block) || itemstack.canPlaceOn(block))
                    }
            }
        }

        result = result && movingObjectPositionIn.typeOfHit == MovingObjectType.BLOCK

        return result
    }

    private fun interpolateAxis(bb: AxisAlignedBB): AxisAlignedBB =
        AxisAlignedBB(
            bb.minX - mc.renderManager.viewerPosX,
            bb.minY - mc.renderManager.viewerPosY,
            bb.minZ - mc.renderManager.viewerPosZ,
            bb.maxX - mc.renderManager.viewerPosX,
            bb.maxY - mc.renderManager.viewerPosY,
            bb.maxZ - mc.renderManager.viewerPosZ,
        )
}
